Cleanup everything

endgame opened the pull request:
Closes #5.

~~This is WIP, so don't merge yet, but it's at the point where I want to get eyes on it.~~

~~It's still crashing a bit, and sockets with pending writes can lose data. I need to fix those things before merge, but I hope you'll agree that the code is simpler and clearer.~~

endgame committed:
Fix Reflex.Backend.Socket.Connect

* Try every returned AddrInfo from getAddrInfo
* Do the connection in a background thread
* Simplify interface

endgame committed:
Tidy client stuff

endgame committed:
Remove dep on reflex-binary

endgame committed:
Use a NonEmpty list for AddrInfos

endgame committed:
Clean up Accept

endgame committed:
Make Accept's fields mimic Socket's in terms of order and name

endgame committed:
Clean up Socket:

* Actually closes sockets
* Fires sClosed event exactly once per disconnection
* Radically simpler

endgame committed:
Stop sockets from instantly closing

endgame committed:
Turn on SO_REUSEADDR for listen sockets

endgame committed:
Bump library version

endgame committed:
Depend on microlens-th instead of lens

dalaing commented on reflex-backend-socket.cabal:
```diff
@@ -19,52 +19,51 @@ library
+                     , containers        >= 0.5     && < 0.7
+                     , dependent-map     >= 0.2     && < 0.3
+                     , dependent-sum     >= 0.4     && < 0.5
+                     , microlens-th      >= 0.4.2.2 && < 0.5
```
If we're pulling in `reflex`, then we are pulling in `lens` transitively anyhow - see [here](https://github.com/reflex-frp/reflex/blob/develop/reflex.cabal#L51).  So adding `microlens` is probably increasing the dependency footprint.

endgame replied:
Have switched back to `lens`.

endgame committed:
Socket: Fix racing between socket close/write events, and busy loops

endgame committed:
Rename tenses on output events for consistency; documentation

endgame committed:
Use lens instead of microlens-th since reflex uses it anyway

endgame committed:
Fix up examples and make sure they build

endgame committed:
Socket: Remove redundant do

endgame committed:
Fix Others example

endgame commented on src/Reflex/Backend/Socket/Accept.hs:
```diff
@@ -1,123 +1,161 @@
+        acceptLoop sock =
+          let
+            exHandlerAccept :: IOException -> IO (Maybe (Socket, NS.SockAddr))
+            exHandlerAccept e = do
```
It is not clear to me that this is the correct behaviour: if the server socket errors once, should we really stop accepting connections forever?

endgame replied:
I am starting to think this is not what we should do: if we fail to accept because we're using too many open files (say), perhaps the correct behaviour is to keep going? Some of the other open connections will go away, and then we can resume normal service.

endgame replied:
I tried a variant of the server code where errors were signaled without terminating the accept loop, and it ran out of files and hard-locked my system. I think it's okay for us to terminate the accept socket on the first error. A library client who wants a robust server can then build in their own delays, and try to bring the listen socket back up, or whatever.

endgame commented on src/Reflex/Backend/Socket.hs:
```diff
@@ -1,192 +1,190 @@
+     )
+  => SocketConfig t
+  -> m (Socket t)
+socket (SocketConfig sock maxRx eTx eClose) = do
```
I am a lot more confident in this now, but I think it needs someone else to think _hard_ about it as well.

endgame commented:
Using the version of refex-backend-socket in this branch, I can crash my telnet server if I open/close too many sockets and do nothing with the `_aError` event on the accepting socket. In such a scenario, there's no way to get new events into the FRP network (all clients are dead, and the server socket closed itself in response to the error), so having reflex seize up and die seems fine?

endgame committed:
Only queue ByteStrings when socket is Open

endgame committed:
Do not block when socket is Draining and payloadQueue is empty

endgame committed:
Use align instead of merging events by hand

endgame commented on src/Reflex/Backend/Socket/Accept.hs:
```diff
@@ -1,123 +1,161 @@
+                NS.close sock
+                onClosed ()
+              Just () -> do
+                (Just <$> NS.accept sock) `catch` exHandlerAccept
```
this isn't quite right: if I am blocked on accept, and my `_acClose` fires, I won't notice until I get unblocked, either from an IOException or because someone tried to connect to me.

endgame replied:
Non-blocking accept, or throw an exception at this thread in the `performEvent_ $ _acClose`...?

endgame replied:
Correct way is to close the socket from the other thread, and catch the error thrown by `accept`, apparently.

endgame commented on src/Reflex/Backend/Socket/Accept.hs:
```diff
@@ -1,123 +1,161 @@
+data Accept t = Accept
+  { _aAcceptSocket :: Event t (Socket, NS.SockAddr)
+    -- ^ A new connection was accepted, including its remote address.
+  , _aClose :: Event t ()
```
Allow passing extra socket options?

endgame commented on src/Reflex/Backend/Socket/Accept.hs:
```diff
@@ -1,123 +1,161 @@
+  , _aError :: Event t IOException
+    -- ^ An exception occurred. Treat the socket as closed after you
+    -- see this. If the socket was open, you will see the
+    -- '_aListenClosed' event fire as well, but not necessarily in the
```
Wrong name

endgame commented on src/Reflex/Backend/Socket/Accept.hs:
```diff
@@ -1,123 +1,161 @@
+     )
+  => AcceptConfig t
+  -> m (Accept t)
+accept (AcceptConfig mHost mService listenQueue eClose) = do
```
-> m (Event t (Either (NonEmpty (AddrInfo, IOException)) (Accept t)))

endgame commented on src/Reflex/Backend/Socket/Accept.hs:
```diff
@@ -1,123 +1,161 @@
+    -- event fires or if the socket closes in response to a caught
+    -- exception caught.
+  , _aError :: Event t IOException
+    -- ^ An exception occurred. Treat the socket as closed after you
```
This will be no longer true, once we fix exception handling inside acceptloop.

endgame replied:
We're going to retain the current behaviour that an exception inside acceptLoop kills the listen socket.

endgame commented on src/Reflex/Backend/Socket.hs:
```diff
@@ -1,192 +1,189 @@
+    queueSend bs = STM.readTVar state >>= \case
+      Open -> STM.writeTQueue payloadQueue bs
+      _ -> pure ()
+    queueClose = STM.writeTVar state Draining
```
Set state to `Draining` only if state is `Open`.

endgame reviewed (commented):
Found more things to fix after review with @LightAndLight .

endgame committed:
Separate errors from binding and errors from accepting

endgame committed:
Only put a socket into 'Draining' if it was previously 'Open'

endgame committed:
Tidy error handling in accept socket

endgame committed:
Reflex.Backend.Socket.Accept: Unblock the accepting thread on close

We tell the accepting thread to finish up by pulling its socket out of
a TMVar and closing it out from underneath it.

endgame committed:
Add shutdown command to example server

endgame committed:
Doc fixups for Accept

endgame committed:
Allow sets of socket options (Accept)

endgame committed:
Haddock fixups

endgame commented:
@dalaing I fixed all the things I found with @LightAndLight . I think it's ready for you to have a go. I recommend coming at it with a combative mindset: think "what is wrong with this code?", as I've been tripped up by subtle bugs for several days straight `:-)`

dalaing commented on src/Reflex/Backend/Socket.hs:
```diff
@@ -1,192 +1,191 @@
+              mBytes <- atomically $
+                STM.readTVar state >>= \case
+                  Closed -> pure Nothing
+                  Draining -> STM.tryReadTQueue payloadQueue
```
Does this need a `retry` like in the case for `Open`?

I'm thinking about what happens if there are multiple firings of the close event, which could cause contention on `state`, which if you were really unlucky might mean you are in the `Draining` state with things in the `payloadQueue` and `mBytes` is `Nothing` and so you close before the queue is properly drained.

I need to stare at this for longer to properly work out if this is an issue.

dalaing commented on src/Reflex/Backend/Socket.hs:
```diff
@@ -1,192 +1,191 @@
+                    >>= maybe STM.retry (pure . Just)
+
+              case mBytes of
+                Nothing -> onClosed ()
```
This should close the socket on its way out.

dalaing replied:
Alternative: when `mBytes` is Nothing, do a state transition to `Closed`, and have a third thread whose job is to watch for that transition, and to close the socket and call `onClosed` when the transition occurs.

endgame replied:
Doing this via a sentinel thread adds the nice property that every branch of the {tx,rx}Loops end in either `loop` or `shutdown`.

dalaing commented on src/Reflex/Backend/Socket/Connect.hs:
```diff
@@ -1,91 +1,70 @@
+    in do
+      res <- runExceptT $ do
+        addrs <- getAddrs
+        let attempts = withExceptT Last . tryConnect <$> addrs
```
Should this be consistent with the error reporting in `Accept`?

endgame replied:
I've changed this, but now I have two isomorphic error types and no good idea of what a unifying name might be (apart from `Either`). Thoughts?

endgame replied:
I called it `SetupError` and moved it into its own module.

endgame committed:
Actually close sockets when done

endgame committed:
Reflex.Backend.Socket.Connect.connect: return errors like 'accept'

endgame committed:
Unify error types

endgame committed:
Hlint pass

endgame committed:
Haddock nit

endgame committed:
Use LambdaCase

endgame commented:
@dalaing thanks for your help. Consider whether this gives you new ideas for #1.
